### DISCUSSION TITLE
How to set the correct Jupyter kernel for the generated notebook

### DISCUSSION THREAD

**Discussion by rgaiacs** [quarto-dev/quarto-cli#9013] [Q&A] [answered]
### Description

# Use case

I'm preparing a lesson in `.qmd` that uses R and I want to export it to `.ipynb` for my students to load on the Jupyter Lab environment they have access on the university's Jupyter Hub. When student open the `.ipynb`, the R kernel should be active kernel.

# Minimal Working Example

Given `mwe.qmd` as

``````
---
title: "MWE"
format:
  ipynb: default
---

Lorem ipsum.

```{r}
1 + 1
```
``````

`quarto render` will run successfully:

```
processing file: mwe.qmd
1/2                  
2/2 [unnamed-chunk-1]
output file: mwe.knit.md

pandoc 
  to: ipynb
  output-file: mwe.ipynb
  standalone: true
  default-image-extension: png
  
metadata
  ipynb-title-block: /opt/quarto/share/formats/ipynb/templates/title-block.md
  title: MWE
  
Output created: mwe.ipynb
```

The produced `mwe.ipynb` is

```
{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# MWE\n",
        "\n",
        "Lorem ipsum."
      ],
      "id": "a96f4d14-3d66-429c-a550-5db284516e78"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[1] 2"
          ]
        }
      ],
      "source": [
        "1 + 1"
      ],
      "id": "76cce413-d8ca-4f4e-a59f-c63d83c03275"
    }
  ],
  "nbformat": 4,
  "nbformat_minor": 5,
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    }
  }
}
```

# Question

What changes should I do on my `.qmd` to have the `.ipynb` with the R kernel instead of the python3 kernel?

<hr />

Edited based on https://github.com/quarto-dev/quarto-cli/discussions/9013#discussioncomment-8719677

`quarto render` for

``````
---
title: "MWE"
jupyter:
  kernelspec:
    display_name: R 4.3.1 (xr)
    language: R
    name: xr
format:
  ipynb: default
---

Lorem ipsum.

```{r}
1 + 1
```
``````

(with the [`xeus-r`](https://github.com/jupyter-xeus/xeus-r) installed) will output

```
Starting xr kernel...Done

Executing 'mwe.quarto_ipynb'
  Cell 1/1: ''...Done

pandoc 
  to: ipynb
  output-file: mwe.ipynb
  standalone: true
  default-image-extension: png
  
metadata
  ipynb-title-block: /opt/quarto/share/formats/ipynb/templates/title-block.md
  title: MWE
  jupyter:
    kernelspec:
      display_name: R 4.3.1 (xr)
      language: R
      name: xr
  
Output created: mwe.ipynb
```

and `mwe.ipynb` is

```
{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# MWE\n",
        "\n",
        "Lorem ipsum."
      ],
      "id": "56d8998d-696b-4e78-8f10-50d84afd95e4"
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [
        {
          "output_type": "display_data",
          "metadata": {},
          "data": {
            "text/plain": [
              "[1] 2"
            ]
          }
        }
      ],
      "source": [
        "1 + 1"
      ],
      "id": "75e0846e"
    }
  ],
  "nbformat": 4,
  "nbformat_minor": 5,
  "metadata": {
    "kernelspec": {
      "name": "xr",
      "display_name": "R 4.3.1 (xr)",
      "language": "R"
    },
    "language_info": {
      "name": "R",
      "codemirror_mode": "",
      "file_extension": "R",
      "mimetype": "text/x-R",
      "nbconvert_exporter": "",
      "pygments_lexer": "",
      "version": "4.3.1"
    }
  }
}
```

as expected.

# Maybe a Feature Request

Quarto has a heuristic to decided the engine to use. This heuristic could also define the Jupyter kernel to be used in the generated `.ipynb`.

**Comment by mcanouil** [comment, marked as answer]
See <https://quarto.org/docs/computations/python.html#kernel-selection>.

> This heuristic could also define the Jupyter kernel to be used in the generated

Very unlikely to happen considering the number of kernels and their various way of being installed, their various versions, etc.
See for example: https://blog.jupyter.org/meet-xeus-r-a-future-proof-jupyter-kernel-for-r-1adc5fdd09ab?gi=1722538e69d9
If you want a particular engine/kernel, then make it explicit by setting it.

The "heuristic" is very very simple: if R code cell then `engine: knitr` else `engine: jupyter`.
Note that to run R code from a Jupyter/Python you gonna need some dependencies, e.g., `IRKernel`.

Additionally, there is no guarantee the kernel you will set will be available to your students under the same name, version, etc.
Selecting a kernel in a Jupyter Notebook, is as simple as clicking on the dropdown menu and selecting the appropriate value.

**Reply by mcanouil** [reply]
FYI: I just edited a bit my answer.

**Reply by rgaiacs** [reply]
Dear @mcanouil,

would be possible for `quarto render` produce a output like

```
processing file: mwe.qmd
1/2                  
2/2 [unnamed-chunk-1]
output file: mwe.knit.md

pandoc 
  to: ipynb
  output-file: mwe.ipynb
  standalone: true
  default-image-extension: png
  
metadata
  ipynb-title-block: /opt/quarto/share/formats/ipynb/templates/title-block.md
  title: MWE
  
Output created: mwe.ipynb

WARNING: The document was executed using knitr but the produced mwe.ipynb is configured to load the Python Jupyter kernel. You might want to specify the Jupyter Kernel in your document. Read more details at https://quarto.org/docs/computations/execution-options.html#engine-binding
```

Thanks!

**Reply by mcanouil** [reply]
Not sure because what if you were using Julia or any kernel.
Why only show a message when engine is knitr?

_(anyhow not really up to me)_

**Reply by rgaiacs** [reply]
> Not sure because what if you were using Julia or any kernel.

So, does the `.ipynb` produced by Quarto always have the kernel set to Python unless the author explicit provides the `kernelspec`?

**Reply by mcanouil** [reply]
I believe so.

**Reply by cderv** [reply]
> So, does the .ipynb produced by Quarto always have the kernel set to Python unless the author explicit provides the kernelspec?

The difference here is that if you provide R code cells, and kernel spec, then **knitr** will be used. Then converting to `ipynb` is done by pandoc conversion, and I believe it will assume python kernel by default. 

Whereas, when the kernelspec is passed, it will render using Jupyter engine, not **knitr**. this probably makes a difference where the conversion keeps the kernel used in the input. 

I don't know exactly how that works, and we could be done. But in the case where knitr engine is used, there is no way to know which R engine to use. There is not a default one really.

> Why only show a message when engine is knitr?

I understand that this problem will really be when converting to `ipynb` format from a .qmd using a non jupyter engine. I don't think there is way to specify which engine to use in the output. Interesting limitation... 🤔  I'll look into it more closely as I wonder who (quarto or pandoc) is keeping the kernel really; 

Anyhow, seems you got the right solution here for now
